src/evtx_chunk.rs

```diff
@@ -250,32 +253,36 @@ impl<'a> Iterator for IterChunkRecords<'a> {
         );
 
         let mut tokens = vec![];
-        let iter = match deserializer.iter_tokens(Some(binxml_data_size)) {
+        let iter = match deserializer.iter_tokens(Some(binxml_data_size)).context(
+            err::FailedToDeserializeRecord {
+                record_id: record_header.event_record_id,
+            },
+        ) {
             Ok(iter) => iter,
-            Err(e) => return Some(Err(format_err!("{}", e))),
+            Err(err_ctx) => return Some(Err(err_ctx.error)),
         };
 
         for token in iter {
-            match token {
+            match token.eager_context(err::FailedToDeserializeRecord {
+                record_id: record_header.event_record_id,
+            }) {
                 Ok(token) => {
                     trace!("successfully read {:?}", token);
                     tokens.push(token)
                 }
-                Err(e) => {
-                    error!("Tried to read an invalid token!, {}", e);
-
-                    if log::log_enabled!(Level::Debug) {
-                        let mut cursor = Cursor::new(self.chunk.data);
-                        cursor
-                            .seek(SeekFrom::Start(
-                                e.offset().expect("Err to have offset information"),
-                            ))
-                            .unwrap();
-                        dump_cursor(&cursor, 10);
-                    }
+                Err(err) => {
+                    //                    if log::log_enabled!(Level::Debug) {
+                    //                        let mut cursor = Cursor::new(self.chunk.data);
+                    //                        cursor
+                    //                            .seek(SeekFrom::Start(
+                    //                                e.offset().expect("Err to have offset information"),
+                    //                            ))
+                    //                            .unwrap();
+                    //                        dump_cursor(&cursor, 10);
+                    //                    }
 
                     self.offset_from_chunk_start += u64::from(record_header.data_size);
-                    return Some(Err(e.into()));
+                    return Some(Err(err));
```

Comment:
err.fail()?

Reply:
At this point the err is already an Error.

`fail` is used on context selectors to, which cannot automatically be promoted to errors.
```
impl<I> UserIdInvalid<I>
where
    I: Into<i32>,
{
    fn fail<T>(self) -> Result<T, Error> { /* ... */ }
}
```

https://docs.rs/snafu/0.3.1/snafu/guide/the_macro/index.html#context-selectors